lib/LaTeXML/Core/Document.pm

```diff
@@ -1821,9 +1873,21 @@ sub wrapNodes {
     $self->setNodeBox($new, $box); }
   foreach my $node (@nodes) {
     $new->appendChild($node); }
-  $self->afterClose($new);
+  $self->afterClose($new) unless $leave_open;
   return $new; }
 
+# Check if $node, or any of it's children, are the current node, and thus still "open"
+# Maybe a better way, such as explicitly marking _open ?
```

Comment:
this is interesting for XML reasons. The recursion could get expensive if someone wraps a big subtree with a new parent, while some of its terminal leaf nodes is still open.

I would even be tempted to suggest avoiding the checks altogether and requiring a separate method for firing the `afterClose`. If we think about unsuspecting web developers using a `wrapNodes` method, we would only expect a re-parenting of a set of nodes, without extra side-effects. The method is a little tricky to think about (for me) as it is. Maybe even changing it to:
```perl
wrapNodes($self, $parent_node, @nodes_to_reparent)
```

would be safer, while adding a higher-level method which could work with a qname, e.g.
```perl
# fires ->afterClose
wrapClosedNodes($self, $qname, @nodes)

# does not fire ->afterClose
wrapOpenNodes($self, $qname, @nodes)
```

Recursing to check if open every time may get expensive, looking at where it is currently used:
```
lib/LaTeXML/Core/Rewrite.pm
187:    $node = $document->wrapNodes('ltx:XMWrap', @nodes); }
204:    my $wrapper = $document->wrapNodes('ltx:XMWrap', @nodes);
206:    $node = $document->wrapNodes('ltx:XMDual', $wrapper);

lib/LaTeXML/Core/Alignment.pm
590:  $document->wrapNodes('ltx:thead', @heads) if @heads;
591:  $document->wrapNodes('ltx:tbody', @rows)  if @rows;
592:  $document->wrapNodes('ltx:tfoot', @foots) if @foots;

lib/LaTeXML/Core/Document.pm
431:        my $text = $self->wrapNodes($elementname, $child);
2406:=item C<< $node = $document->wrapNodes($qname,@nodes); >>

lib/LaTeXML/Package/TeX.pool.ltxml
2018:      push(@newnodes, $document->wrapNodes('ltx:p', @n)); }
3388:        $text = $document->wrapNodes('ltx:text', $text); }
3477:            $document->wrapNodes('ltx:XMText', $m); }
```

To my intuition it may even be better to require from to the binding developer to call `->afterClose` explicitly, removing it from the current method, compared to deeply recursing in the children for every call...

Reply:
Concern noted, but one motivation for the "fully automatic" was *exactly* the poor ole unsuspecting developer wondering why the close-node daemon was sometimes running and sometimes not. While it may be expensive to sniff it out from within `Document`, it's hard to figure out at the binding level whether it's needed or not, and then to get it done at the right time.

I'd rather improve how the check is done: The suggested alternative was off-hand; probably marking `_closed` would be safer than `_open`, since it would be hard to keep track of the latter, but with the former, once closed it's done.  Still, it would require more testing & exploration than fits in this PR; might be better to come back to this topic separately with some appropriate benchmarks.

Another aside: I do like the built-in automation here. And with the box management, I was hoping that a further evolution could possibly clean up a lot of unreadable code used to manage MathForks and such. But that's for another time, as well.